pkg/roachprod/cloud/cluster_cloud.go

```diff
@@ -346,14 +355,46 @@ func CreateCluster(l *logger.Logger, opts []*ClusterCreateOpts) error {
 			p = (p + 1) % providerCount
 		}
 
+		var vms vm.List
+		var vmsLock syncutil.Mutex
+		// Create VMs in parallel across all providers.
+		// Each provider will return the list of VMs it created, and we append
+		// them to the cached Cluster.
 		if err := vm.ProvidersParallel(o.CreateOpts.VMProviders, func(p vm.Provider) error {
-			return p.Create(l, vmLocations[p.Name()], o.CreateOpts, o.ProviderOptsContainer[p.Name()])
+			vmList, err := p.Create(l, vmLocations[p.Name()], o.CreateOpts, o.ProviderOptsContainer[p.Name()])
+			if err != nil {
+				return err
+			}
+			vmsLock.Lock()
+			defer vmsLock.Unlock()
+			vms = append(vms, vmList...)
+			return nil
 		}); err != nil {
-			return err
+			return c, err
+		}
+
+		c.VMs = append(c.VMs, vms...)
+	}
+
+	// Check that we did create the expected number of VMs
+	if len(c.VMs) != nodesCreated {
+		return c, errors.Errorf("Created %d VMs, while %d nodes were expected", len(c.VMs), nodesCreated)
```

Comment:
What would be an example behind this mismatch?

Reply:
Fair question, each provider should report an error if they weren't able to create the VM, so I guess this error should never be raised. Since I had the data, it felt like an extra check was a good idea, but it's probably useless.